### PR TITLE
rename Molecule > IP Tokens

### PR DESCRIPTION
**Pull request by elmariachi111**

### CODE REVIEW

**Comment by benjileibo** [comment]
This looks perfect, thank you for tagging me

File: script/dev/CrowdSale.s.sol [hunk 1]
```diff
@@ -118,7 +118,7 @@ contract FixtureCrowdSale is CommonScript {
         placeBid(alice, 600 ether, saleId, abi.encodePacked(r, s, v));
         (v, r, s) = vm.sign(charliePk, ECDSA.toEthSignedMessageHash(abi.encodePacked(terms)));
         placeBid(charlie, 200 ether, saleId, abi.encodePacked(r, s, v));
-        console.log("LOCKED_MOLECULES_ADDRESS=%s", address(lockedMolToken));
+        console.log("LOCKED_IPTS_ADDRESS=%s", address(lockedMolToken));
```
Comment (DorianWilhelm) [inline script/dev/CrowdSale.s.sol:121]: lockedIPToken

File: script/prod/RolloutV23.sol [hunk 2]
```diff
@@ -5,8 +5,8 @@ import "forge-std/Script.sol";
 import "forge-std/console.sol";
 import { ERC1967Proxy } from "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
 import { IPNFT } from "../../src/IPNFT.sol";
-import { IPermissioner, TermsAcceptedPermissioner } from "../../src/Permissioner.sol";
-import { Synthesizer } from "../../src/Synthesizer.sol";
+import { IPermissioner, TermsAcceptedPermissioner } from "../../src/helpers/test-upgrades/SynthPermissioner.sol";
+import { Synthesizer } from "../../src/helpers/test-upgrades/Synthesizer.sol";
```
Comment (DorianWilhelm) [inline script/prod/RolloutV23.sol:9]: Tokenizer

File: src/Permissioner.sol [hunk 3]
```diff
@@ -35,34 +35,34 @@ contract TermsAcceptedPermissioner is IPermissioner {
...
         if (!isValidSignature(tokenContract, _for, signature)) {
             revert InvalidSignature();
         }
         emit TermsAccepted(address(tokenContract), _for, signature);
     }
 
     /**
-     * @notice checks whether `signer`'s `signature` of `specificTermsV1` on `moleculeId` is valid
+     * @notice checks whether `signer`'s `signature` of `specificTermsV1` on `tokenContract.metadata.ipnftId` is valid
      * @param tokenContract Molecules
```
Comment (DorianWilhelm) [inline src/Permissioner.sol:56]: IPToken

File: src/Permissioner.sol [hunk 4]
```diff
@@ -78,7 +78,7 @@ contract TermsAcceptedPermissioner is IPermissioner {
      * @notice this yields the message text that claimers must present as signed message to burn their molecules and claim shares
      * @param tokenContract ISynthesizedToken
```
Comment (DorianWilhelm) [inline src/Permissioner.sol:79]: IPToken

File: src/Tokenizer.sol [hunk 5]
```diff
@@ -5,27 +5,18 @@ import { UUPSUpgradeable } from "@openzeppelin/contracts-upgradeable/proxy/utils
...
+/// @title Tokenizer 1.1
 /// @author molecule.to
-/// @notice synthesizes an IPNFT to an ERC20 token (called molecules) and controls its supply.
-///         Allows molecule holders to withdraw sales shares when the IPNFT is sold
-contract SynthesizerNext is UUPSUpgradeable, OwnableUpgradeable {
-    event MoleculesCreated(
+/// @notice tokenizes an IPNFT to an ERC20 token (called IPT) and controls its supply.
+contract Tokenizer is UUPSUpgradeable, OwnableUpgradeable {
+    event TokensCreated(
         uint256 indexed moleculesId,
```
Comment (DorianWilhelm) [inline src/Tokenizer.sol:20]: ipTokenId maybe?

File: src/Tokenizer.sol [hunk 6]
```diff
@@ -38,17 +29,17 @@ contract SynthesizerNext is UUPSUpgradeable, OwnableUpgradeable {
 
     IPNFT internal ipnft;
 
-    mapping(uint256 => MoleculesNext) public synthesized;
+    //this is the old term to keep the storage layout intact
+    mapping(uint256 => IPToken) public synthesized;
```
Comment (DorianWilhelm) [inline src/Tokenizer.sol:33]: aaah, this is annoying that we cannot change this :/

File: src/helpers/test-upgrades/SynthPermissioner.sol [hunk 7]
```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: MIT
```
Comment (DorianWilhelm) [inline src/helpers/test-upgrades/SynthPermissioner.sol:1]: This is necessary to upgrade the old Synthesizer for which we cannot change the storage layout, correct?
Can we add this as a comment here?

File: subgraph/makeAbis.sh [hunk 8]
```diff
@@ -4,9 +4,69 @@ cat ../out/IERC20Metadata.sol/IERC20Metadata.json | jq .abi > abis/IERC20Metadat
...
-cat ../out/Molecules.sol/Molecules.json | jq .abi > ./abis/Molecules.json
+cat ../out/IPToken.sol/IPToken.json | jq .abi > ./abis/IPToken.json
 cat ../out/TimelockedToken.sol/TimelockedToken.json | jq .abi > ./abis/TimelockedToken.json
 cat ../out/SalesShareDistributor.sol/SalesShareDistributor.json | jq .abi > ./abis/SharedSalesDistributor.json
 cat ../out/Permissioner.sol/TermsAcceptedPermissioner.json | jq .abi > ./abis/TermsAcceptedPermissioner.json
 cat ../out/StakedLockingCrowdSale.sol/StakedLockingCrowdSale.json | jq .abi > ./abis/StakedLockingCrowdSale.json
+
+cat ../out/Tokenizer.sol/Tokenizer.json | jq .abi > ./abis/_Tokenizer.json
+
+jq '. +=  [{
```
Comment (DorianWilhelm) [inline subgraph/makeAbis.sh:15]: Can we add a comment here why this stunt is needed?

File: subgraph/schema.graphql [hunk 9]
```diff
@@ -6,36 +6,36 @@ type Ipnft @entity {
...
   lockedToken: TimelockedToken #the locked token that bidders partially receive when claiming
-  molecules: [Molecule!] @derivedFrom(field: "reactedIpnft")
-  crowdsales: [CrowdSale!] @derivedFrom(field: "reactedIpnft")
+  iptBalance: [IPTBalance!] @derivedFrom(field: "ipt")
+  crowdsales: [CrowdSale!] @derivedFrom(field: "ipt")
   createdAt: BigInt!
 }
 
-type Molecule @entity {
+type IPTBalance @entity {
```
Comment (DorianWilhelm) [inline subgraph/schema.graphql:32]: I like it! Probably the best name so for

File: subgraph/src/termsAcceptedPermissionerMapping.ts [hunk 10]
```diff
@@ -1,25 +1,25 @@
...
-  let molecule = Molecule.load(moleculesId)
+  let iptBalance = IPTBalance.load(balanceId)
 
-  if (!molecule) {
-    let reacted = ReactedIpnft.load(event.params.tokenContract.toHexString())
+  if (!iptBalance) {
+    let reacted = IPT.load(event.params.tokenContract.toHexString())
     if (!reacted) {
-      log.warning('molecules {} not found for signature', [moleculesId])
+      log.warning('molecules {} not found for signature', [balanceId])
```
Comment (DorianWilhelm) [inline subgraph/src/termsAcceptedPermissionerMapping.ts:16]: ipt

File: test/CrowdSalePermissioned.t.sol [hunk 11]
```diff
@@ -43,7 +43,7 @@ contract CrowdSalePermissionedTest is Test {
         (bidder, bidderPk) = makeAddrAndKey("bidder");
         vm.startPrank(deployer);
 
-        auctionToken = new Molecules();
+        auctionToken = new IPToken();
         auctionToken.initialize("MOLECULES", "MOL-0001", Metadata(42, msg.sender, "ipfs://abcde"));
```
Comment (DorianWilhelm) [inline test/CrowdSalePermissioned.t.sol:47]: IPToken, IPT-0001

File: script/prod/RolloutV23.sol (hunk 2 above)
Comment (elmariachi111) [reply script/prod/RolloutV23.sol:9]: this is the V23 rollout that already happened

File: test/CrowdSalePermissioned.t.sol (hunk 11 above)
Comment (elmariachi111) [reply test/CrowdSalePermissioned.t.sol:47]: it also works fine like this :) and it's a nice nod to the past

**Comment by benjileibo** [comment]
Bravo 🥹🥹

**Pull request merged**